presto-main/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql;
+
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.analyzer.MaterializedViewColumnMappingExtractor;
+import com.facebook.presto.sql.tree.CreateMaterializedView;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.Query;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.metadata.MetadataUtil.toSchemaTableName;
+import static com.facebook.presto.spi.ConnectorMaterializedViewDefinition.TableColumn;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public class MaterializedViewUtils
+{
+    private MaterializedViewUtils() {}
+
+    /**
+     * Compute the 1-to-N column mapping from a materialized view to its base tables.
```

Comment:
Sorry. I miss some context. So have to ask some high level questions. Could you please briefly describe how do we use this column mapping? Is it used in the write path to detect which materialized view to update or in the read path to find candidate materialized views?

Mapping a join output column to a join input column is not always semantically correct. For example, the outer side columns could be null values which are not in the original table. We need to make sure it doesn't cause correctness issues for the use cases of this column mapping. (I know we only support INNER JOIN for now. But are we planning to support outer joins?)

A join output column could also map to an expression of multiple input columns or a field inside a flattened array.  What does the mapping look like in these cases?

Reply:
Hi @yuanzhanhku , yeah, a bit of the context. The column mapping is needed both in the write path for refresh MV and in the read path for partition stitching of select from MV. More details on the logic with examples could be found in this design doc: read path (https://fb.quip.com/ZdrfAF313Mco#TeBACArnMea), write path (https://fb.quip.com/ZdrfAF313Mco#TeBACA2Sniy).

Right now we enforce strong restrictions on the partition mapping and the query shape, aimed to make sure we can get the read logic and refresh logic correct. First, the partition mapping has to be 1-1, allowing only direct select or alias. Second, only inner equal-join is supported. More on the restriction, https://fb.quip.com/ZdrfAF313Mco#TeBACApMp5Z

Right now we support only inner join but not outer join, solely because it is not easy to get the logic correct even w/ the 1-1 partition mapping restriction. And of course we'd need to handle the INNER JOIN use cases first before getting to OUTER JOIN use cases.

Reply:
Thanks for the explanation and providing the pointers.

IIUC, lots of the complexity (such as enforcing partition mapping, extracting the column mappings) are added because we are modeling the materialized view as a global materialized view concept which is basically a different table with its own partitioning. That is why we have to add the logic to make sure its partitioning matches the base tables' partitioning. 
Have we considered the local materialized view model which makes the MV local to a partition. i.e., the materialized views always have the same partitioning as the base tables, and the MV query is defined in the partition level. This should simplify lots of things, e.g., no need to enforce partition mapping. Partition stitching will also be straightforward. In that case, we may not need to extract the column mapping any more?

Reply:
The idea of local MV sounds interesting to me! Would like to learn more about how it works and use cases. One of the problems we look to solve by MV is to speed up queries w/ high aggregation, for dashboard performance. Can local MVs help with aggregation queries?

Reply:
Yes, it does. Basically, you can consider each partition of a table as an individual table and support any MV query patterns over it, including aggregations, unions and joins.
When there are multiple tables involved in the MV, we just need to make sure that all of them have the same partitioning. And the MV is defined locally for each of their corresponding partition tuple.

Global MVs are more general which allows you to do rollups across partitions. But with your 1:1 partition matching limitation, you are basically limiting the global MV use cases to only the ones supported by local MVs.

We can discuss more about this. No need to address it in this PR.